### PR TITLE
Ticks for GeoAxis, other things

### PR DESCRIPTION
**Pull request by asinghvi17**
This PR aims to do the following things:

- [X] Make GeoMakie work on Mac M1
- [X] Implement working ticks and tick labels for GeoMakie
- [x] Implement a "direction finder" for GeoMakie ticklabels
- [X] Automatically find valid limits for projection spaces (import from #100)
- [x] Tie GeoAxis into the layout system with protrusions, etc.

There are also some new examples and general cleanup (with a little bit of mess added for the future :D).

This PR depends on https://github.com/JuliaGeo/Proj4.jl/pull/57, so it will take some time before it can be merged.  (This is simply because only Proj4.jl/pull/57 works on the new Mac now).

Out of curiosity, @visr, would you happen to have an approximate target for the rename of Proj4.jl?

In order to use this branch locally, you will need to `]add Makie#master MakieCore#master CairoMakie#master GLMakie#master`.

Solves #104 #99 #77 #55 #32

### CODE REVIEW

**Comment by asinghvi17** [comment]
These are the basic surface and scatter examples:
surf_example
scatter_example

As you can see, there are still some problems with ticks.  I'm not sure if it's scaling or they're not getting updated properly.

Tests will fail until Proj is released, which is why this is marked as a draft PR.

**Comment by asinghvi17** [comment]
This is the projection example from #102 
projections
and an orthographic projection with correct limits
```julia

fig = Figure()
ga = GeoAxis(
    fig[1, 1],
    dest="+proj=ortho",
    limits = (-90,90,-90,90), # IMPORTANT - you MUST have proper limits!
    coastlines = true,
    title = "Orthographic projection with proper limits"
)
hidedecorations!(ga)
sp = surface!(ga, lons, lats, field; shading = false, colormap = (:plasma, 0.45))
cb = Colorbar(fig[1, 2], sp)
save("orthographic_with_limits.png", fig, px_per_unit=2)
fig
```
orthographic_with_limits

File: src/geoaxis.jl [hunk 1]
```diff
@@ -55,74 +58,282 @@ display(fig)
...
+                Ref(camera(scene)),
+                # go from dataspace (transformation applied to inputspace) to pixelspace
+                Ref(:data), Ref(:pixel),
+                # apply the transform to go from inputspace to dataspace
+                Makie.apply_transform(
+                    scene.transformation.transform_func[],
+                    Point2f.(_xtickvalues, ylimits[][1])
+                )
+            )
+        ) .+ Ref(Point2f(pxarea.origin))
```
Comment (asinghvi17) [inline src/geoaxis.jl:184]: @SimonDanisch I would appreciate some help on this if you have the time - basically, I am trying to project the tick coordinates (which I get in input (lat/long) space from `get_ticks`), to dataspace and then back to `ax.scene`'s pixel space.  Then, I move them to `ax.blockscene`'s pixel space by adding `pixelarea(ax.scene)[].origin`.

I'm not sure if this entire lift statement is just not triggering when it's supposed to, or if something else is going on, but it looks like the ticks are somehow placed at an old position?

Comment (asinghvi17) [reply src/geoaxis.jl]: Nevermind, found the issue - I didn't update the ticklabel positions, assuming they would be read when the ticklabels updated, but I guess that wasn't the case :(

**Comment by asinghvi17** [comment]
At this point, this is the state of the PR:

<details>
    <summary>Basic example - surface
</summary>

```julia
using GeoMakie, CairoMakie

lons = -180:180
lats = -90:90
# Create some field of values across `lons` and `lats`
# This grid can be of any density, but note that the
# time it takes to plot scales with the grid size!
field = [exp(cosd(l)) + 3(y/90) for l in lons, y in lats]

# Surface example
fig = Figure()
ax = GeoAxis(fig[1,1])
surface!(ax, lons, lats, field; shading = false)
save("surf_example.png", fig; px_per_unit=2)
display(fig)
```
surf_example

</details>

<details>
    <summary>Basic example - scatter
</summary>

```julia
slons = rand(lons, 2000)
slats = rand(lats, 2000)
sfield = [exp(cosd(l)) + 3(y/90) for (l,y) in zip(slons, slats)]

fig = Figure()
ax = GeoAxis(fig[1,1])
scatter!(slons, slats; color = sfield)
save("scatter_example.png", fig; px_per_unit=2)
display(fig)
```

scatter_example

</details>

<details>
<summary>Projection example</summary>

```julia
lons = -180:180
lats = -90:90
field = [exp(cosd(l)) + 3(y / 90) for l in lons, y in lats]

fig = Figure()
ax1 = GeoAxis(fig[1, 1], dest = "+proj=vitk1 +lat_1=45 +lat_2=55",
    coastlines = true, title = "vitk1")
ax2 = GeoAxis(fig[1, 2], dest = "+proj=wintri",
    coastlines = true, title = "wintri")

surface!(ax1, lons, lats, field; shading = false, colormap = (:plasma, 0.45))
surface!(ax2, lons, lats, field; shading = false)
hidedecorations!(ax1)
save("projections.png", fig; resolution = (900, 400), px_per_unit=2)
fig
```

projections

</details>
GeoMakie can find the input domain of your transformation through `GeoMakie.find_transform_limits(p::Proj.Transformation) in utils.jl` assuming that it takes lat-long coordinates as input.  An example is below:

```julia
fig = Figure()
ga = GeoAxis(
    fig[1, 1],
    dest="+proj=ortho",
    lonlims = Makie.automatic,
    coastlines = true,
    title = "Orthographic projection with proper limits"
)
hidedecorations!(ga)
sp = surface!(ga, lons, lats, field; shading = false, colormap = :rainbow_bgyrm_35_85_c69_n256)
cb = Colorbar(fig[1, 2], sp)
save("orthographic_with_limits.png", fig, px_per_unit=2)
fig
```
orthographic_with_limits

Here I used `lonlims=Makie.automatic` to trigger automatic limit setting.  You can do that to the `latlims` parameter as well.

Note that the color does not go to blue in the image, but the colormap does - this indicates that the surface is working correctly.


A consequence of these changes is that the API has changed.  If you want to specify ticks, you can just do `ga.xticks[] = ...` like you would for a normal plot, and it will accept LinearTicks, WilkinsonTicks, MultipleTicks or even just a range as in the previous examples.

I'm not sure if this autolimit thing should be the standard yet, for one reason - it cannot get the limit to be exactly 90 or 180 (since it is an approximation).  Therefore if you're plotting the whole globe, GeoAxis will end up ignoring lon=±180, and lat=±90.

**Comment by visr** [comment]
> Out of curiosity, @visr, would you happen to have an approximate target for the rename of Proj4.jl?

Let's try to make that happen in a week or two. Any input there is welcome.

**Comment by asinghvi17** [comment]
New examples!  Check out this video of a rotating Earth:

https://user-images.githubusercontent.com/32143268/165003843-db5984f0-9ccf-49f7-847e-88fd63e80bb4.mp4

and this lovely orthographic projection of Italy's states, updated from #45 
italy

Just to note: rendering the rotating Earth (360 frames at 60 fps) takes ~20 minutes using CairoMakie :D

**Comment by asinghvi17** [comment]
Note: documentation is still broken since CairoMakie is not precompiling.

**Comment by asinghvi17** [comment]
I changed the logic of the ticks quite a bit - it's a little complex now, but with the nonlinearities involved in the plots it's the only solution.

Basically, once I have a set of tick positions and labels, I compute the "normal" vector pointing directly out of the plot at each location, and the bounding box of each label.  Then, I shift the label's position by that normal vector multiplied by the diagonal of its bounding box (which I also calculate).  This means that ticks are always outside the axis, which was not the case before.

Additionally, I have written some logic which removes overlapping tick labels.  Currently it's pretty brute force, but it only takes ~1ms for a reasonable number of ticks, which I feel is OK.

Basically, what this does is compute the distances between all unique pairs of tick points, and if any pair is overlapping, it will delete the first tick only.  This means that in an orthographic projection, there will be no x-ticks; the y-ticks always take primacy.  This is easy enough to change, but there had to be _some_ default so I figured this was as good as any.

If CI passes now, I will likely merge the PR - it works with Proj4.jl as is, but a couple of one-line changes in Project.toml and it will work with Proj.jl after the rename.  I held off from changing the rectangle transformation for this reason.

New instructions to use the PR:
```julia
]add GeoMakie#as/geoaxis Makie#master CairoMakie#master MakieCore#master
```
and enjoy!

**Comment by asinghvi17** [comment]
Some images for the README:
geoax1
geoax2
geoaxes

**Comment by Datseris** [comment]
Hi @asinghvi17 , long time no see, happy to see you here :) I am back to work next week, currently recovering from an eye surgery. I would like to test this PR in my office where I do my meteorology work. If you can wait a week before merging, I can give you a proper PR review!

**Comment by asinghvi17** [comment]
No worries - I'll probably wait to merge this until Proj4 is renamed anyway.  It will be good to get an opinion on the API from people who use this kind of stuff regularly.

> recovering from an eye surgery

Oof, hope you get well soon!

**Comment by asinghvi17** [comment]
Last todo: I need to have the axis update code trigger on all possible observables that could influence it.

File: README.md [hunk 2]
```diff
@@ -1,15 +1,71 @@
...
-[![Build Status](https://gitlab.com/JuliaGPU/GeoMakie-jl/badges/master/pipeline.svg)](https://gitlab.com/JuliaGPU/GeoMakie-jl/pipelines)
-
-## Installation
+![CI](https://github.com/JuliaPlots/GeoMakie.jl/actions/workflows/ci.yml/badge.svg)
 
-This package is **in development** and will **break often**.  You can install it from the REPL like so:
+This package is **in development**, and may break often.  You can install it from the REPL like so:
 ```julia
 ]add GeoMakie
 ```
```
Comment (lazarusA) [inline README.md:13]:
```suggestion
In order to start using `GeoMakie` make sure to also add a backend (GLMakie, CairoMakie, WGLMakie), e.g.,
```julia
julia>]
pkg> add GLMakie
```
You may check the installed versions with:
```Julia
julia>]
pkg> st GLMakie GeoMakie
```
Start using the package:
```julia
using GLMakie, GeoMakie
```
```

File: README.md [hunk 3]
```diff
@@ -1,15 +1,71 @@
...
 [![Dev](https://img.shields.io/badge/docs-dev-blue.svg)](https://JuliaPlots.github.io/GeoMakie.jl/dev)
-[![Build Status](https://gitlab.com/JuliaGPU/GeoMakie-jl/badges/master/pipeline.svg)](https://gitlab.com/JuliaGPU/GeoMakie-jl/pipelines)
-
-## Installation
+![CI](https://github.com/JuliaPlots/GeoMakie.jl/actions/workflows/ci.yml/badge.svg)
 
-This package is **in development** and will **break often**.  You can install it from the REPL like so:
+This package is **in development**, and may break often.  You can install it from the REPL like so:
 ```julia
 ]add GeoMakie
```
Comment (lazarusA) [inline README.md:11]:
```suggestion
julia>]
pkg> add GeoMakie
```

Comment (asinghvi17) [reply README.md:11]: IMO this is a pretty common theme in READMEs to have `]add PackageName` with nothing else...plus it's easily copy-pasteable while the multiline version is not.

**Comment by asinghvi17** [comment]
Leaving some notes for when I come back:

Tick label alignment was still being set by the axis, but I have overridden that.  The issue is now that the tick deflection vector is off, and in different ways depending on which quadrant of the hemisphere going from (-90,90) lon it is in.  The code is here:

https://github.com/JuliaPlots/GeoMakie.jl/blob/9dbdbbf47c229df8c03cb9b2b9bb8b5c179a3a1f/src/utils.jl#L195-L226

basically, I need a smarter system to select which direction to "push" the ticklabel vectors in.  

These are some images of what I have right now:
it
iTerm2 t3u2ky
OSM_GeoMakie
sc
usa

It's clear that there are some regressions, but overall the new features seem to be working well.

File: README.md (hunk 3 above)
Comment (lazarusA) [reply README.md:11]: sure. It's just that in my experience teaching Julia, I always need to repeat myself, specially at the beginning, that we need to go first into the package mode and then we can add a new package. That's it. The one liner is also ok, ;)

**Comment by asinghvi17** [comment]
I put a quick script together to visualize what the tick padding algorithm was doing.  It seems as though I am only computing some sort of local curvature vector, which I should not be doing.  Though it works well enough on the large scale, it seems to perform badly on the small scale.  The red arrows in the plot below represent the direction and magnitude of the tick padding vector, although the magnitude is doubled for easier viewing.

ticklabel_directions

**Comment by asinghvi17** [comment]
I just looked at Cartopy's code...they seem to basically try padding until it works.
https://github.com/SciTools/cartopy/blob/63fd1e3ce7bc146a0e1a5b120441dabd4d38651c/lib/cartopy/mpl/gridliner.py#L980-L1000

Is there a way we can find whether a `Rect` and a line intersect?  If so I could implement this pretty easily.  It would have the advantage of the tick labels being a little more accurate in position as well.

**Comment by Datseris** [comment]
I can test this starting tomorrow or so, but I wanted to download stuff first. @asinghvi17 do I really need master versions of e.g., CairoMakie? Why?

**Comment by asinghvi17** [comment]
Mostly it's because Makie master is subtly breaking (some internal APIs have changed) and I wanted to build for that instead of having to rewrite code again later.  Plus, there was a refactor of the `Axis` code which makes things easier on the GeoAxis end, along with some bugfixes in CairoMakie.

**Comment by asinghvi17** [comment]
@Datseris quick comment - all you need to do is add `Makie#master MakieCore#master CairoMakie#master GLMakie#master`.  I had written that you need a new branch of Proj4.jl in the description, but that's not required anymore.

**Comment by asinghvi17** [comment]
Hacky but good: nonlinear clip in CairoMakie!

iTerm2 5vcrvC

**Comment by Datseris** [comment]
I'll be testing this today and tomorrow! Will look at existing docs for code snippets (but of course will be primarily using the code snippets of my work)

**Comment by asinghvi17** [comment]
Cool!  Ping me if you have any questions (or if something doesn't build right, we recently merged a lot of PRs to Makie) @Datseris

**Comment by Datseris** [comment]
@asinghvi17 I am testing this now. Here are some comments:

- In the `GeoAxis` docstring you say "Note that if your data is not adjusted to the WGS84 datum" and the sentence ends there abruptly. Not sure what is supposed to follow.
- Personally, I strongly dislike having E W N S in my ticks and prefer to have a minus `-` in front. I'd really prefer a way to do this for the ticks, and hopefully a way to also set it as a default project-wide.
- How does one set specific ticks here? I see that `lonlims` and `latlims` are given to the axis, but If I want to have ticks exactly at latitude 20 and 30, how to do so?
- It is not clear to me what `line_density` does or stands for.
- Setting `ax.xticks = LinearTicks(3)` completely breaks the plot (e.g., the plot coming from the `contourf` example), so maybe we should be advertising in the docstring to access `ax.xticks`.
- My data are in the format of having longitude in 0 to 360 degrees. This doesn't seem to work. I used the argument `lonlims = (0, 360)`, but I get: 

```
Error showing value of type Figure:
ERROR: MethodError: no method matching gl_convert(::Tuple{Int64, Int64})
Closest candidates are:
```

**Comment by Datseris** [comment]
Apologies, I was wrong. Setting `xticks` works as expected, its just that I did it with the axis that had longitude 0 to 360 that made nonsensical results. E.g.

image

**Comment by visr** [comment]
I don't really have much preference for +/- or E-W-N-S tick labels. +/- takes up less space, which can be tight when the meridians come together, giving a cleaner look. E-W-N-S is probably more traditional.

**Comment by asinghvi17** [comment]
Huh.  I tried that with `dest="+proj=eqearth +lon_0=180", lonlims = (0, prevfloat(360f0))` and I get this:
</img>
but plotting a normal contourf yields 
</img>

which seems to indicate that there is still some wrapping. If I change `lons` to, e.g., 
```julia
lons = Float32.(collect(0:360))
lons[end] = prevfloat(lons[end])
```
then I get 
</img>
so I guess you just have to force Proj to avoid wrapping.

**Comment by asinghvi17** [comment]
@Datseris the issue is that the last value in `lons` is 360, which Proj translates to `0`.  To get around that I set the last value to `prevfloat(360f0)`, which means that Proj transforms it to the right end of the plot as opposed to the left end.

The replacement is:
```julia
using GeoMakie, CairoMakie

lons = Float32.(collect(0:360))
lons[end] = prevfloat(lons[end])
lats = -90:90
field = [exp(cosd(l)) + 3(y/90) for l in lons, y in lats]

fig = Figure()
ax = GeoAxis(fig[1,1]; dest = "+proj=eqearth +lon_0=180", lonlims = (0, prevfloat(360f0)))
contourf!(ax, lons, lats, field; shading = false)

fig

```

**Comment by asinghvi17** [comment]
By the way, interactivity should work now, but I don't have any applications to test it with.  If you do, could you try it out and let me know?  I believe that the issue previously was that the inverse transform either didn't exist or didn't invert properly (off by a factor of 100_000), both of which should now be fixed.

**Comment by Datseris** [comment]
i have some and i'll test it tomorrow

**Comment by Datseris** [comment]
Alright, tested interactive stuff:

- Plotting observables and updating them works as expected!
- Click+drag, which would normally zoom in an normal axis, is broken. So I disabled it using `MakieLayout.deactivate_interaction!(ax, :rectanglezoom)`.
- `poin = select_point(ax.scene)` also works!!! I am very happy about that! It gives me the point in longitude latitude coordinates as expected! However, the point I select, and where my cursor is, are not the same. In the figure I post below, the point selection is plotted to the top left of my cursor position (not very off, but off by about 20 degrees or so).

image

**Comment by Datseris** [comment]
(MWE: just do 
```julia
MakieLayout.deactivate_interaction!(ax, :rectanglezoom)
point = select_point(ax.scene)
on(point) do val
    println(val)
end
```
after the contourf example)

**Comment by asinghvi17** [comment]
I played around with the interaction a little bit, but it looks like the mouseposition function is seriously off.  I will investigate further next week.

**Comment by asinghvi17** [comment]
https://github.com/JuliaPlots/Makie.jl/blob/070f2a73bd53623f38593a269cf04c729b8c5a7a/src/interaction/interactive_api.jl#L212-L223

https://github.com/JuliaPlots/Makie.jl/blob/6c0f7c758a1446aa8fa98c5074f40be215b58bc2/src/camera/projection_math.jl#L241-L279

This is the code which `select_point` calls - I will try to roll my own mouse selection function, and if it works for this will PR it to Makie.

**Comment by lazarusA** [comment]
any thoughts one merging this soon ? ... it looks ok... still some corner cases, but good enough to be use on a daily basis.

**Comment by Datseris** [comment]
I agree. The `select_point` interaction isn't really a priority for me either!

**Comment by asinghvi17** [comment]
Sounds good - the whole projection mess should probably wait for Proj's big refactor anyway.  The only other work I want to do currently is fixing the tick mess and slightly refactoring GeoAxis to plot its grid to the block's scene instead of the Axis scene.  That can definitely wait though.

The other project which would be really good to do, is refactoring GeoAxis to be its own Block - instead of a thin wrapper around Axis as it is now.  The only new thing that one would need to do is to figure out a way to keep the size correct!

A new GeoAxis could be made based off of LScene, with the appropriate axis-like attributes and other similar handling.  This would really un-complicate a lot of things and remove the potential for weird bugs if the Axis code changes slightly.
